### ISSUE TITLE
The single shared large-layer slot forces a full embedding-matrix H2D every step on an untied checkpoint (~33 ms for 500 MiB, measured at 16 GB/s)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#975]
## What was measured

`benchmarks/harness/layer0_wait.py` (committed harness, re-run 2026-09-14
15:53 from the committed path after an earlier run was found to have come
from a session-scratchpad copy — see "Where it comes from" below), driving
the probe's own instruments and labelling each load bracket with the layer it
belongs to. Cold synthetic Llama-70B shape, NF4, depth (`read_ahead`) 2, 3
steps, no warm-up, instrumented from step 0. JSON:
`layer0_wait_cold_synth70b_nf4.json`. Absolute step times (49.2, 39.4, 39.3 s)
are explicitly NOT throughput evidence per the record (no warm-up, instrumented
throughout); the ratios within each step are what this measures.

**Per-load brackets** (record's table, §8):

| step | layer 0, forward / backward | other layers, mean / median / max | vocabulary loads | brackets |
|---|---|---|---|---|
| 0 | 17 / 271 ms | 252 / 283 / 420 ms | 662, 386 ms | 40.6 s of 49.2 s = 82.6% |
| 1 | 17 / 259 ms | 214 / 266 / 533 ms | 33, 22 ms | 33.5 s of 39.4 s = 84.9% |
| 2 | 16 / 268 ms | 213 / 270 / 327 ms | 33, 22 ms | 33.3 s of 39.3 s = 84.8% |

**Per-step compute-stream stalls, from the JSON's `stalls` array** (Finding
8a's table):

| step | total stall | `large:model.embed_tokens.weight` | `layer000` forward | mean, other 79 decoder layers |
|---|---|---|---|---|
| 0 | 33.17 ms | 32.66 ms | 0.17 ms | 0.0021 ms |
| 1 | 47.92 ms | 32.61 ms | 14.99 ms | 0.0020 ms |
| 2 | 47.58 ms | 32.91 ms | 14.35 ms | 0.0020 ms |

Record's own reading (Finding 8a, quoted): "The embedding out-stalls layer 0
in every one of the three steps, layer 0 is second in every one of them, and
the other 79 decoder layers (158 stall entries) sit at 0.002 ms — four orders
of magnitude below the embedding in all three steps (15,700x / 16,300x /
16,200x), and three to four below layer 0 in steps 1 and 2 (7,500x / 7,100x;
in step 0 layer 0 barely stalls, so that gap is only 84x). In steps 1 and 2
the embedding's stall is nearly its entire load — 32.61 of 33.12 ms, 32.91 of
33.46 ms — i.e. the compute thread waits out the step's first read almost
exactly, which is the predicted single-member-group behaviour and not an
inference about it. Step 0 is the exception in the useful direction: the
reader has the build phase to start that 662 ms first touch, so only
32.66 ms of it reaches the compute thread."

**Finding 8** (layer 0 itself, for contrast): "layer 0's forward load is the
FASTEST load of the step, 16-17 ms against a 213-252 ms mean, because the
reader has the whole build and embedding phase to stage it before the compute
thread asks for it."

**Materiality, stated in the record's own words** (§8, ruling on Finding 8a):
"The step's entire stall — the embedding, `lm_head`, layer 0 and all 79 other
decoder layers together, every one of the 162 entries — is 33/48/48 ms, which
is 0.07-0.12% of the step (0.067% / 0.122% / 0.121%), against per-load read
brackets that are 82.6-84.9% of the same steps. Refusing to plan out of a
single-member group would buy back at most that."

## Where it comes from

`benchmarks/gate-971-async-nvme-source.md`, section 8 ("The step-head stall is
the embed fetch, not layer 0 — and neither is material"), Finding 8 and
Finding 8a, and section 10 ("Verdict"). Hardware: RTX 5070 Laptop GPU
(Blackwell sm_120), Windows 11 Pro, driver 616.92, Python 3.12.10, torch
2.14.0+cu130. Branch `feat/async-nvme-source` at `9e5ce63a`. This is a re-run:
the record notes the table published first came from a scratchpad copy of the
driver (its JSON recorded `"driver": "C:\\...\\scratchpad\\layer0_wait.py"`
and carried no `source_class` key) and was superseded 2026-09-14 15:53 by a
run from the committed `benchmarks/harness/layer0_wait.py`, which always
writes `source_class`; the new JSON records
`"driver": "benchmarks/harness/layer0_wait.py"` and
`"source_class": "AsyncDiskSource"`. Every figure above is from that re-run.

The record's stated cause: "`_plan_queue` walks one index out of a
single-member group, so a vocabulary-sized read is planned ahead of decoder
layer 0 at the moment the compute thread blocks on layer 0" — and separately,
Finding 8a: "`_plan_queue` walks one index out of a single-member group, so
the embedding is the step's first read with nothing staged ahead of it."
`_plan_queue` is defined at `src/soup_cli/utils/async_disk_source.py:446`.

## What was NOT measured

- **Any fixture other than the synthetic 70B shape at depth 2.** The record
  states this is "scoped to one fixture, one depth, one shape" (§8, closing
  Finding 8).
- **A fix implementation.** The record considered and rejected acting on
  this: "The addendum offered a follow-up (refuse to plan out of a
  single-member group); on this evidence it is not needed" (§8, Finding 8).
- **Any warm-fixture or 7B-scale version of this stall pattern** — §8 is
  cold-70B only.
- **Whether priming the embed fetch one step ahead (rather than refusing to
  plan out of a single-member group) changes anything**, since neither was
  implemented or measured.
- **Impact at other batch sizes / sequence lengths** — only batch 1 x seq 512
  cold-70B was run through this harness.

## Proposed fix path

This is explicitly NOT a throughput fix — the record's own materiality
statement (0.07-0.12% of step time against 82.6-84.9% spent in reads overall)
means this issue is latency hygiene at the step boundary, not a throughput
regression. Say so in any PR description; do not frame it as recovering
meaningful step time.

Candidate directions, either compatible with the record's own suggestion:
1. In `_plan_queue` (`src/soup_cli/utils/async_disk_source.py:446`), refuse
   to plan a read out of a single-member group as the very first queued item
   at the head of a step — i.e. detect when the walk would place a
   vocabulary-sized (large) group ahead of the next decoder layer the
   compute thread is about to block on, and defer it.
2. Alternatively, prime the embedding fetch one step ahead of when it's
   needed (during the tail of the previous step), so it is already resident
   by the time `_plan_queue` would otherwise schedule it at the step head.

Either change touches `_plan_queue` / `_claim_slot` / `get` in
`src/soup_cli/utils/async_disk_source.py`, which per this record's own
standing constraint (§1, §6) means the race battery
(`.claude/probes/2026-09-13-async-nvme/`, read_ahead 1/2/4/8 x 3) must be
re-run clean before the change counts as verified — this is a local,
gitignored probe directory, not part of the public record, but the
constraint is inherited from the same work this record measures.

## Acceptance criteria

- Re-run `benchmarks/harness/layer0_wait.py` (committed path) on the same
  cold-70B / depth-2 fixture after the fix; the embedding's per-step stall
  (currently 32.6-32.9 ms, the dominant entry in Finding 8a's table) should
  drop out of the step head without introducing a new demand-miss stall
  elsewhere in the 162-entry `stalls` array.
- Race battery (read_ahead 1/2/4/8 x 3) clean, per the standing constraint on
  changes to `_plan_queue` / `_claim_slot` / `get`.
- The PR description states the expected magnitude up front (currently
  ~33-48 ms out of a 30-48 s cold step, i.e. ~0.1%) so reviewers don't read
  it as a throughput fix.

## Links

- #971
- #973
- `benchmarks/gate-971-async-nvme-source.md`

**Comment by VanshSharmaPES** [comment]
I'd like to take this one. I'll start by separating host read time from host-to-device copy time in the head-of-step stall, so the fix targets the right cause, and I'll re-run the race battery before proposing any change to `_plan_queue`.

**Comment by VanshSharmaPES** [comment]
Traced the mechanism rather than acting on the `_plan_queue` theory directly, and I think the root cause is different from what's stated.

**What the two brackets actually measure**

`stream_probe.py`'s `Instruments` wraps two things per large-layer load:
- `copies` (`load_async`): a CUDA-event bracket on the copy stream around `source.get()` + `dst.copy_(..., non_blocking=True)`. Since the copy stream executes in program order, `end.record(stream)` only fires once the H2D copy actually completes — so this is real device-copy time, not CPU dispatch overhead.
- `stalls` (`wait`): brackets `torch.cuda.current_stream().wait_event(...)` on the compute stream. This measures how much of that copy's duration wasn't hidden by other GPU work before the compute stream needed it.

**Why the embedding pays this every step regardless of `_plan_queue`**

`LargeLayerBufferPool` (`layer_stream_runtime.py:901`) holds `embed_tokens` and the untied `lm_head` in **one shared buffer** — by design, to avoid keeping both vocabulary-sized tensors resident. `load_async` only skips the copy `if self.owner == key`. Given the actual call sequence per step:
1. `tail_prefetch` fires at the last decoder layer of the forward pass and loads `lm_head` into that slot (owner becomes `lm_head`).
2. The next step's `_prime` forward-pre-hook loads `embed_tokens` into the *same* slot — owner mismatches, so this is an unconditional fresh H2D copy, issued right as forward starts, with almost nothing else on the GPU to overlap it against.

That matches the numbers in the record far better than a disk-read ordering bug would: a genuinely fixed-size H2D copy of one matrix, re-issued every step because the single slot just swapped away from it, with copy-to-use latency close to zero because of when in the step it's issued. That's consistent with Finding 8a's own observation that "the embedding's stall is nearly its entire load" every step.

**Implication for `_plan_queue`**

Refusing to plan a read out of a single-member group (or priming it a step ahead through `AsyncDiskSource`) wouldn't touch this: `source.get()` isn't even the bottleneck once the shard is warm — steps 1 and 2 already show the vocabulary read itself is fast (22-33 ms total, and per the record that's read *and* copy combined at the `AsyncDiskSource` layer). The dominant cost is the H2D transfer forced by the shared-slot ping-pong between `embed_tokens` and `lm_head`, which has nothing to do with disk scheduling.

**Where I'd look for an actual fix**

Mirror the existing forward-tail prefetch, but for the embedding: instead of loading `embed_tokens` in `_prime` (right when forward starts and needs it immediately), issue that `load_async` at the tail of the *backward* pass of the current step, while the last few decoder layers are still recomputing — giving the copy real time to overlap with compute before the next step's forward asks for it. That's a change to `layer_stream_runtime.py` (`StreamPrefetcher`/`_prime`/backward hook wiring), not to `AsyncDiskSource._plan_queue` / `_claim_slot` / `get`, so it should sidestep the race-battery constraint this issue's proposed fix would have triggered.

**Why I haven't touched code yet**

I don't have hardware that can run the acceptance criteria — the synthetic cold-70B fixture needs real VRAM headroom and the original record used an RTX 5070 Laptop GPU; what I have here is a 4GB RTX 3050, which can't load this fixture. I also don't have the gitignored race-battery probe directory (`.claude/probes/2026-09-13-async-nvme/`) locally. Rather than land an unverified change against those constraints, I wanted to get the redirected diagnosis in front of whoever can run `layer0_wait.py` on the 5070 — happy to write the backward-tail-prefetch change and a probe to confirm the shared-slot copy is the actual cost, if someone can run it, or to hand this off if hardware access is the blocker.

**Comment by VanshSharmaPES** [comment]
Correction + supporting math for my previous comment.

**Correction:** I mischaracterized the `copies` bracket (the `load_async` CUDA-event pair). I wrote that it measures "real device-copy time, not CPU dispatch overhead" — that's wrong. It wraps `source.get()` and `dst.copy_` together, and `layer0_wait.py`'s own docstring says outright that this bracket "includes the CPU-side `source.get()`." Disregard that sentence. It doesn't change the underlying claim, which rests on the separate `wait()` bracket (the `stalls` table), not `copies`: by the time `wait()` runs, `load_async()`'s `source.get()` call has already returned, so `wait()`'s elapsed time is purely the remaining async H2D copy, independent of `_plan_queue` / disk timing.

**Supporting numbers I should have included the first time:**
- `tie_word_embeddings: False` is hardcoded in the synthetic fixture generator (`synth_checkpoint.py:117`), with a separate `lm_head.weight` written out (`synth_checkpoint.py:216`) — confirms this fixture is genuinely untied, so `LargeLayerBufferPool`'s single shared slot really does ping-pong between `embed_tokens` and `lm_head` every step, not just in theory.
- `embed_tokens`/`lm_head` are stored as bf16 in this fixture (`synth_checkpoint.py:118,203`), not NF4 — only the decoder layers are quantized. With vocab≈32000 ("v32k") and hidden=8192, that's a 32000×8192×2-byte ≈ 524 MB matrix. A 524 MB H2D copy at a laptop-class ~16 GB/s sustained PCIe throughput is ≈33 ms, which lines up with the observed 32.6–32.9 ms almost exactly.

That's a fairly direct quantitative match for "this is a real, largely un-hideable H2D transfer forced by the shared slot," rather than a `_plan_queue` disk-ordering effect.

**Comment by MakazhanAlpamys** [comment]
Thank you for the correction — and specifically for correcting a sentence that nobody had challenged. Retracting "real device-copy time, not CPU dispatch overhead" when the docstring says the bracket includes `source.get()` is the difference between a measurement and a claim, and it cost you nothing to leave it standing. I would rather read that than a tidy thread.

**The arithmetic checks out, and I re-derived it rather than agreeing with it:**

```
32000 x 8192 x 2 bytes = 524,288,000 B = 524.3 MB (500.0 MiB exactly)

at 25 GB/s -> 20.97 ms       at 16 GB/s -> 32.77 ms
at 20 GB/s -> 26.21 ms       at 12 GB/s -> 43.69 ms

implied bandwidth at the observed 32.75 ms: 16.01 GB/s
```

16 GB/s is the right order for a **pinned** H2D on a laptop PCIe 4.0 x16 link, and decisively not the ~6 GB/s a pageable copy would give. So the number is not just consistent with your explanation, it is consistent with the specific mechanism — a pinned async copy of the full embedding matrix — rather than with anything disk- or dispatch-shaped. That is a stronger result than the one you originally claimed.

And your two fixture facts are what make it load-bearing: `tie_word_embeddings: False` hardcoded at `synth_checkpoint.py:117` with a separate `lm_head.weight` at `:216`, and both stored bf16 at `:118,203` while only the decoder layers are quantised. So the single shared slot in `LargeLayerBufferPool` really does ping-pong between `embed_tokens` and `lm_head` every step in this fixture, and each swap is a genuine 500 MiB transfer that no amount of prefetch ordering can hide.

**Which means the issue's title is now wrong, and that is worth saying out loud.** It says `_plan_queue` stalls the compute thread on the vocabulary fetch. Your own evidence says the stall is not a `_plan_queue` ordering effect at all — it is an un-hideable H2D forced by the pool geometry. Same observation, different cause, and the fix would be in a different place. I will retitle it.

**One connection you could not have seen, because it landed today.** PR #1061 adds a second head-sized allocation for an untied streamed head, and in reviewing it I found it falsifies a documented invariant of `predicted_peak_bytes` — that function's docstring promises it "never under-predicts", and it still charges **one** slot (`stream_setup.py:893`, `large_layer_bytes=large_buffer_bytes`). On Llama-3.1-8B that gap is `128256 x 4096 x 2 = 1.05 GiB`.

Put next to your finding, those are two halves of the same design question: **one slot costs a full embedding-matrix H2D on every step; two slots cost roughly a GiB of VRAM and remove it.** On an 8 GB card — the cards layer streaming exists for — that is a real trade rather than an obvious win, and nobody has measured which side is better. If you want it, that is a much more interesting issue than the one you opened, and your harness is already the instrument for it: the `wait()` bracket you are reading is exactly what would go to zero.

No obligation — you found the thing, so you get first refusal. Either way, thank you for the rigour.

**Comment by VanshSharmaPES** [comment]
Thanks — I'll take it. Filing a follow-up issue now that keeps the two things you found separate, since they need different verification:

1. The `predicted_peak_bytes` undercount — I can verify this one without special hardware, it's static arithmetic against a documented contract, not a throughput measurement.
2. The one-slot-vs-two-slot design question from this issue's stall — does need real hardware and the harness you pointed at.

One correction to how PR #1061 gets used there, from reading its actual diff rather than going by the changelog wording: `_needs_a_private_weight()` gates the clone on `torch.is_grad_enabled()` plus "slot shared by more than one key, module isn't an embedding" — it does not check which loss is running. So the extra head-sized clone fires on every forward through an untied streamed `lm_head` while grad is enabled, not only the preference-loss reference pass the changelog motivates it with. That makes the `predicted_peak_bytes` gap wider than "preference losses on untied checkpoints" once #1061 lands — it's any training step on an untied streamed checkpoint. Filing the issue now, before it merges.
